module/mempool/herocache/backdata/heropool/pool_test.go

```diff
@@ -215,6 +231,97 @@ func TestInvalidateEntity(t *testing.T) {
 	}
 }
 
+// TestAddAndRemoveEntities checks health of heroPool for scenario where entitites are stored and removed in a predetermined order.
+// LRUEjection, NoEjection and RandomEjection are tested. RandomEjection doesn't allow to provide a final state of the pool to check.
+func TestAddAndRemoveEntities(t *testing.T) {
+	for _, tc := range []struct {
+		limit             uint32              // capacity of the pool
+		entityCount       uint32              // total entities to be stored
+		ejectionMode      EjectionMode        // ejection mode
+		operationsIndexes []OperationAndIndex // operation to perform on an entity stored under the given index,
+		// where index ranges from 0 to entityCount
+		finalEntitiesInThePool []uint // indexes of entities residing in the pool at the end of the test
+	}{
+		{
+			limit:                  2,
+			entityCount:            5,
+			ejectionMode:           LRUEjection,
+			operationsIndexes:      []OperationAndIndex{{Add, 0}, {Add, 0}, {Add, 1}, {Add, 2}, {Add, 3}, {Remove, 2}, {Add, 2}, {Remove, 2}},
+			finalEntitiesInThePool: []uint{3},
+		},
+		{
+			limit:                  2,
+			entityCount:            5,
+			ejectionMode:           NoEjection,
+			operationsIndexes:      []OperationAndIndex{{Add, 0}, {Add, 1}, {Add, 2}, {Add, 3}, {Remove, 0}, {Remove, 1}, {Add, 4}, {Add, 3}},
+			finalEntitiesInThePool: []uint{3, 4},
+		},
+		{
+			limit:                  5,
+			entityCount:            6,
+			ejectionMode:           RandomEjection,
+			operationsIndexes:      []OperationAndIndex{{Add, 0}, {Add, 1}, {Remove, 1}, {Add, 2}, {Remove, 0}, {Add, 5}, {Add, 4}, {Add, 1}, {Add, 3}, {Add, 0}},
+			finalEntitiesInThePool: nil,
+		},
+	} {
+		t.Run(fmt.Sprintf("%d-limit-%d-entities", tc.limit, tc.entityCount), func(t *testing.T) {
+			testAddRemoveEntities(t, tc.limit, tc.entityCount, tc.ejectionMode, tc.operationsIndexes, tc.finalEntitiesInThePool)
+		})
+	}
+}
+
+// testAddRemoveEntities allows to add or remove entities in an order given by operationsIndexes. Each OperationAndIndex consists of an operaton to perform
+// on an entity stored under a corresponding index. Index range from 0 to entityCount.
+func testAddRemoveEntities(t *testing.T, limit uint32, entityCount uint32, ejectionMode EjectionMode, operationsAndIndexes []OperationAndIndex, finalEntitiesInThePool []uint) {
+
+	pool := NewHeroPool(limit, ejectionMode)
+
+	entities := unittest.EntityListFixture(uint(entityCount))
+
+	// this map maintains entities currently stored in the pool
+	var insertedEntities map[flow.Identifier]EIndex
+	insertedEntities = make(map[flow.Identifier]EIndex)
+
+	for _, operationAndIndex := range operationsAndIndexes {
+		operation := operationAndIndex.operation
+		index := operationAndIndex.index
```

Comment:
Please consider renaming this index to `ownerIndex` to be consistent with the `pool.Add` interface. Also, this can be generated randomly as it should not affect the pool operation.

Reply:
Thank you for the clarification. I'm still contemplating the implications of selecting the `ownerIndex` randomly and how it might impact the test's functionality. As it stands, in this test, owner index is aligned with the entity’s index in the pool. However, it appears feasible to make the owner index random, thus fully decoupling it from the entity index in the pool. This approach would align more closely with practical applications.

In general, for testing fundamental data structures, a preference towards larger, randomized test cases is advisable compared to hand-crafted ones. Randomized testing can be more effective in identifying unforeseen issues and ensuring that the data structure is robust under a variety of conditions. It's essential to simulate real-world scenarios as closely as possible to ascertain the integrity and reliability of the data structure.

Reply:
I do this call ```pool.Add(entities[ownerIndex].ID(), entities[ownerIndex], uint64(ownerIndex))``` So, indeed I use entities index in the array produced by ```unittest.EntityListFixture(uint(entityCount))``` as an ```ownerIndex```. Indeed, it doesn't have to be so. I agree a more realistic scenario is where an ```ownerIndex``` is a random number attached to an entity index in the test set up.

Reply:
I've added random ownerIds. Please take a look.